{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6610242b",
   "metadata": {},
   "source": [
    "# Locating Motifs Despite Introns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1f2facd",
   "metadata": {},
   "source": [
    "In “Finding a Shared Motif”, we discussed searching through a database containing multiple genetic strings to find a longest common substring of these strings, which served as a motif shared by the two strings. However, as we saw in “RNA Splicing”, coding regions of DNA are often interspersed by introns that do not code for proteins.\n",
    "\n",
    "We therefore need to locate shared motifs that are separated across exons, which means that the motifs are not required to be contiguous. To model this situation, we need to enlist subsequences."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a9234c8",
   "metadata": {},
   "source": [
    "-> A sequence is include 'introns'. Therefore, when the 2 sequences are not match completely, we need to find the sequece that match the parts of the 2 sequences.\n",
    "\n",
    "ex. ATTG is the common subsequence of the following 3 sequences.  \n",
    "common meanse the sequence contains the same letters in the same order of the match seq.  \n",
    "\n",
    "match seq: ATTG  \n",
    "sequence 1: A T T C G  \n",
    "sequence 2: A C C T T T T G  \n",
    "sequence 3: C C A A T T G  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c78cd943",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(fasta: str) -> dict[str, str]:\n",
    "    sequences = {}\n",
    "    header = None\n",
    "    seq = []\n",
    "    fasta = fasta.splitlines()\n",
    "    for line in fasta:\n",
    "        if line.startswith('>'):\n",
    "            if header is not None:\n",
    "                sequences[header] = ''.join(seq)\n",
    "            header = line[1:]  # Remove '>'\n",
    "            seq = []\n",
    "        else:\n",
    "            seq.append(line)\n",
    "    if header is not None:\n",
    "        sequences[header] = ''.join(seq)  # Add the last sequence\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c1de96db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pick up the index of the candidate of common subsequence\n",
    "from itertools import combinations\n",
    "def pick_index(seq_1:str):\n",
    "    index_list = []\n",
    "    n = len(seq_1)\n",
    "    for i in range(n, 0, -1):\n",
    "        for perm in combinations(range(n), i):\n",
    "            index_list.append(perm)\n",
    "    return index_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "87b2beb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract candidate common subsequence by using the index\n",
    "def extract_candidate_common_subseq(seq_1: str, index_tuple: tuple[int]) -> str:\n",
    "    candidate = []\n",
    "    for i in index_tuple:\n",
    "        candidate.append(seq_1[i])\n",
    "    return ''.join(candidate)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cff9713",
   "metadata": {},
   "source": [
    "when we use `in` for itteration, it to repeatedly call `next()` until it finds a match."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "51c28499",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare the caandidate common subsequence with the other sequence\n",
    "def compare_candidate_with_other_seq(candidate: str, seq_2: str):    \n",
    "    itter_seq_2 = iter(seq_2)\n",
    "    return all(char in itter_seq_2 for char in candidate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7945a444",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(fasta_txt):\n",
    "    fasta_dict = read_fasta(fasta_txt)\n",
    "    seqs = list(fasta_dict.values())\n",
    "    seq_1 = seqs[0]\n",
    "    seq_2 = seqs[1]\n",
    "\n",
    "    index_list = pick_index(seq_1)\n",
    "    for index_tuple in index_list:\n",
    "        candidate = extract_candidate_common_subseq(seq_1, index_tuple)\n",
    "        is_common = compare_candidate_with_other_seq(candidate, seq_2)\n",
    "        if is_common:\n",
    "            print(candidate)\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4f078f91",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_input = \"\"\">Rosalind_23\n",
    "AACCTTGG\n",
    ">Rosalind_64\n",
    "ACACTGTGA\"\"\"\n",
    "\n",
    "# result: AACTGG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "01d50006",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AACTTG\n"
     ]
    }
   ],
   "source": [
    "main(sample_input)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
